src/raw.rs

```diff
@@ -292,6 +303,47 @@ pub fn get_string() -> io::Result<String> {
     }
 }
 
+/// Retrieves a list of file paths from the `CF_HDROP` format.
+///
+/// # Pre-conditions:
+///
+/// * [open()](fn.open.html) has been called.
+pub fn get_file_list() -> io::Result<Vec<PathBuf>> {
+    unsafe {
+        let clipboard_data = GetClipboardData(formats::CF_HDROP).if_null_get_last_error()?;
+        let _data_ptr = GlobalLock(clipboard_data).if_null_get_last_error()?;
+        let gen_err_unlock = || -> io::Error {
+            GlobalUnlock(clipboard_data);
+            io::ErrorKind::Other.into()
+        };
+        let num_files = DragQueryFileW(clipboard_data as HDROP, std::u32::MAX, ptr::null_mut(), 0);
+        let file_names: io::Result<Vec<PathBuf>> = (0..num_files)
+            .into_iter()
+            .map(|file_index| {
+                let required_size =
+                    1 + DragQueryFileW(clipboard_data as HDROP, file_index, ptr::null_mut(), 0)
+                        .if_null_to_error(gen_err_unlock)?;
+                let file_str_buf = {
+                    let mut buffer = Vec::with_capacity(required_size as usize);
+                    DragQueryFileW(
+                        clipboard_data as HDROP,
+                        file_index,
+                        buffer.as_mut_ptr(),
+                        required_size,
+                    ).if_null_to_error(gen_err_unlock)?;
+                    // Set length, remove terminating zero
+                    buffer.set_len(required_size as usize - 1);

```

Comment:
use `pop` instead

Reply:
`set_len` is necessary here since the Vec was just allocated with `Vec::with_capacity(required_size as usize)` but not initialized yet so `DragQueryFileW` can write to the allocated part.

Reply:
Ah I see, my bad, I misunderstood.
But then why you subtract 1?
Wouldn't it contain `\0` at the end?

Reply:
`required_size` is calculated above and is the total size that needs to be allocated, including the zero at the end, because that will get written. By setting the length to `required_size - 1`, you both set the state of the `Vec` correctly and remove that zero.

Reply:
I understand this point, but you actually the actual length would still be `required_size`.
So technically you you should set it as `required_size` and then `pop` zero character.
Instead you set incorrect length, which is likely to not cause leaks as it is single peace of memory, but just to be on safe size I'd suggest to correctly set size and remove zero character

Reply:
It can't cause leaks as long as `DragQueryFileW` works as specified since we allocate enough space for the zero. `Vec` will deallocate all of that when dropped since it owns all of the memory. `set_len` is also used safely according to the [docs](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.set_len) (you are allowed to set it to less than capacity). I actually looked this up before, but can't find the post anymore.

Reply:
Yeah I know, it is safe because buffer is single chunk of memory.

I just feel there is no need to introduce extra _unsafely_ for no reason as we're anyway going to discard the buffer
After all `from_wide` accepts slice and you could just take slice of all characters except zero ;)

Reply:
Hmm, I don't see where we are introducing extra unsafety since we have to use `set_len` at least once. `Vec` itself doesn't care about the allocated but yet unused space, so that one zero that's written at the end is completely ignored.
In any case, I could use your suggestion to set the full length and take a slice instead. Shall I change it?

Reply:
By extra unsafety, i mean using `set_len` since it doesn't benefit in any way, I'd prefer if you use slice. Thanks

Reply:
Oh, do you mean not using `set_len` at all? I don't think that's possible, unless you zero-initialize the whole Vec first. The [set_len docs](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.set_len) even give an example doing pretty much what I do here, just not setting it to one less element.

Reply:
OK, done.